oktapam/resource_database.go

```diff
@@ -0,0 +1,418 @@
+	"github.com/okta/terraform-provider-oktapam/oktapam/client"
+	"github.com/okta/terraform-provider-oktapam/oktapam/client/wrappers"
+	"github.com/okta/terraform-provider-oktapam/oktapam/constants/attributes"
+	"github.com/okta/terraform-provider-oktapam/oktapam/constants/descriptions"
+	"strings"
+)
+
+const (
+	MySqlBasicAuth = "mysql.basic_auth"
+)
+
+var managementConnectionDetails = &schema.Resource{
+	Schema: map[string]*schema.Schema{
+		attributes.Hostname: {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: descriptions.DatabaseHostname,
+		},
+		attributes.Port: {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: descriptions.DatabasePort,
+		},
+		attributes.AuthDetails: {
+			Type:        schema.TypeList,
+			Required:    true,
+			MinItems:    1,
+			MaxItems:    1,
+			Description: descriptions.DatabaseAuthDetails,
+			ForceNew:    false,
+			Elem:        mySQLBasicAuthDetails,
+		},
+	},
+}
+
+var mySQLBasicAuthDetails = &schema.Resource{
+	Schema: map[string]*schema.Schema{
+		attributes.Username: {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: descriptions.UserName,
+		},
+		attributes.Password: {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: descriptions.DatabasePassword,
+			Sensitive:   true,
+		},
+		attributes.Secret: {
+			Type:        schema.TypeString,
+			Description: descriptions.DatabaseSecretID,
+			Computed:    true,
+		},
+	}}
+
+func resourceDatabase() *schema.Resource {
+	return &schema.Resource{
+		Description:   descriptions.ResourceDatabase,
+		CreateContext: resourceDatabaseCreate,
+		ReadContext:   resourceDatabaseRead,
+		UpdateContext: resourceDatabaseUpdate,
+		DeleteContext: resourceDatabaseDelete,
+		Schema: map[string]*schema.Schema{
+			attributes.ID: {
+				Type:     schema.TypeString,
+				Computed: true,
+				// Description is autogenerated
+			},
+			attributes.ResourceGroup: {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: descriptions.ResourceGroupID,
+			},
+			attributes.Project: {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: descriptions.ProjectID,
+			},
+			attributes.CanonicalName: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: descriptions.CanonicalName,
+			},
+			attributes.DatabaseType: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: descriptions.DatabaseType,
+			},
+			attributes.RecipeBook: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: descriptions.RecipeBookID,
+			},
+			attributes.ManagementConnectionDetails: {
+				Type:        schema.TypeList,
+				Required:    true,
+				MinItems:    1,
+				MaxItems:    1,
+				Description: descriptions.ManagementConnectionDetails,
+				Elem:        managementConnectionDetails,
+				ForceNew:    false,
+			},
+			attributes.ManagementConnectionDetailsType: {
```

Comment:
With security policies, I opted to use a different method for cases where you have a polymorphic type.  With this, it would end up being something like: 

```
management_connection_details {
  mysql {
   ....
  }
}
```

It does make the code more complicated, but I like it since it hides some of the API implementation details from the user and reduces chances for misconfiguration.

Reply:
Yeah, I like that.

I was also curious if we'd want to remove some of the nesting of auth details & connection details and have them all under `management_connection_details`.  Is that what you're thinking?

e.g.

```
management_connection_details {
  mysql {
     host: ...
     port: ...
     user: ...
     pass: ...
  }
}
```

vs. separating out username, password, secret?

Reply:
i believe so - the main thing is if all of that stuff is contained within the polymorphic object the API expects, then it should be under `management_connection_details.mysql`.  if it is something that's common to all API requests, regardless of the details type, then it should remain outside of it.

Reply:
noting for historical purposes, we'll be going with something like

```
management_connection_details {
  mysql {
     host: ...
     port: ...
     basic_auth {
          user: ...
          pass: ...
     }
  }
}
```